homeassistant/helpers/template.py

```diff
@@ -1599,9 +1599,13 @@ def has_value(hass: HomeAssistant, entity_id: str) -> bool:
     """Test if an entity has a valid value."""
     state_obj = _get_state(hass, entity_id)
 
-    return state_obj is not None and (
-        state_obj.state not in [STATE_UNAVAILABLE, STATE_UNKNOWN]
-    )
+    if state_obj is not None:
```

Comment:
I would suggest either 
- comment that `_collect_state` is called already in `_get_state` if the state is None
- or call `_collect_state` always

Reply:
When `has_value` is called as a filter `_get_state` returns None and does not `_collect_state`.  That's what this fix is for.  I can call `_collect_state` always, but when `_get_state` returns a state, we will be calling `_collect_state` twice.

Reply:
@edenhaus not sure if you got my previous response notification

Reply:
Don't ping me if I don't responed within one day. We have more than 400 open PRs and it will take some time to review your changes again.
I get all the notifications of this repo including your two above...

Your comment is the other way around. When `_get_state` returns `None`, `_collect_sate` was already called in https://github.com/Petro31/core/blob/063c2842b89f61034eccd8d1140922e2fbea7220/homeassistant/helpers/template.py#L1134

Reply:
Sorry, I was trying to continue the conversation and I wasn't sure if you got a notification.  This has had the smash label on it for quite some time (almost a month), it's very easy to tell there's only 21 other smash PRs.  Historically, smashed PRs go pretty quick, especially seeing there's only about 20 of them.  I've been working on this project for 7 years now with various PRs and this has been the slowest smash PR I've ever dealt with.

I will debug into this again to show you that the code you think is being hit is not being hit when applied as a jinja filter.  I know it may seem like it hits `_collect_state` by just reading the code, however I can tell you it does not.  I know this may seem like I'm arguing, but I'd like to you to understand that I spent quite a bit of time researching this issue.  I'll update you and show you the stack trace when I get home later.

Reply:
Actually, it looks like this PR fixed the issue and we can probably dump this PR.

<https://github.com/home-assistant/core/commit/de73cafc8bba413c0f6d496da7148b604031da8f>

Reply:
We should probably keep the test and dump the fix.

Reply:
Can you test that it is working correctly without the fix? If yes, we can keep only the test :)

Reply:
Yep :)

This was taken using 2023.10.4
![image](https://github.com/home-assistant/core/assets/35082313/ec38fa3e-90f0-4acc-a9ad-abecb3c5a8fe)

![image](https://github.com/home-assistant/core/assets/35082313/38255f35-c9c2-4f58-9213-b24c8594f458)

Reply:
I'll add the test case and run the suit when I get home.  If it passes, I'll update the PR to only include the test.